### PR TITLE
validator: make sure Ingress/RouteGroup doesn't have duplicate hosts

### PR DESCRIPTION
**Pull request by MustafaSaber**
Relates to https://github.com/zalando/skipper/issues/1618

### CODE REVIEW

File: dataclients/kubernetes/definitions/routegroupvalidator.go [hunk 1]
```diff
@@ -108,6 +109,18 @@ func (rgv *RouteGroupValidator) validateBackends(item *RouteGroupItem) error {
 	return errorsJoin(errs...)
 }
 
+func (rgv *RouteGroupValidator) validateDuplicateHosts(item *RouteGroupItem) error {
+	var errs []error
+	uniqueHosts := make(map[string]struct{}, len(item.Spec.Hosts))
+	for _, hosts := range item.Spec.Hosts {
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/definitions/routegroupvalidator.go:115]: hosts -> host

File: dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+apiVersion: zalando.org/v1
```
Comment (AlexanderYastrebov) [inline dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml:1]: I think the test fixture for Ingress is missing

File: cmd/webhook/admission/testdata/ingress/invalid-ingress-with-duplicate-hosts.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{
+    "request": {
```
Comment (AlexanderYastrebov) [inline cmd/webhook/admission/testdata/ingress/invalid-ingress-with-duplicate-hosts.json:2]: I still think we do not need to add tests for admission webhook when we change validator, its just increases size of PR but adds no value as webhook simply calls validator

File: dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml (hunk 2 above)
Comment (MustafaSaber) [reply dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml:1]: Right now we don't validate ingress in skipper we only do in ValidationWebhook, I think we will need another PR to migrate current ingress adhock validation to validator object.

Comment (MustafaSaber) [reply dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml:1]: What I mean by adhock validation https://github.com/zalando/skipper/blob/master/dataclients/kubernetes/ingress.go#L257C1-L269C2

Comment (MustafaSaber) [reply dataclients/kubernetes/definitions/testdata/validation/route-with-duplicate-hosts.yaml:1]: https://github.com/zalando/skipper/pull/2757

**Comment by AlexanderYastrebov** [comment]
@MustafaSaber What's up with this one? Looks like it needs rebase

**Comment by MustafaSaber** [comment]
This was based on `backendValidation` maybe we can roll it forward first and then rebase this?

**Comment by MustafaSaber** [comment]
Revert PR
https://github.com/zalando/skipper/pull/2825

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by MustafaSaber** [comment]
👍

File: dataclients/kubernetes/definitions/ingressvalidator.go [hunk 4]
```diff
@@ -50,3 +51,15 @@ func (igv *IngressV1Validator) validateRoutesAnnotation(annotations map[string]s
...
+
+func (igv *IngressV1Validator) validateHosts(item *IngressV1Item) error {
+	var errs []error
+	uniqueHosts := make(map[string]struct{}, len(item.Spec.Rules))
+	for _, rule := range item.Spec.Rules {
+		if _, ok := uniqueHosts[rule.Host]; ok {
+			errs = append(errs, fmt.Errorf("duplicate host %q", rule.Host))
+		}
+		uniqueHosts[rule.Host] = struct{}{}
+	}
```
[the remark this answers is not in the document]
Comment (MustafaSaber) [reply dataclients/kubernetes/definitions/ingressvalidator.go:63]: We agreed to disable duplicate hostname validation for ingress.

**Pull request merged**